### DISCUSSION TITLE
How to use msw with typescript?

### DISCUSSION THREAD

**Discussion by Toshinaki** [mswjs/msw#1920] [Q&A] [answered]
Last time I used `msw` it's still `rest.xxx` and easy to define request data type and return data type.

Today I started a new project and noticed the api changed to `http`.

Take [the code](https://mswjs.io/docs/network-behavior/rest/#reading-path-parameters) from the docs for example, 
there're 2 return types here, 
`return new HttpResponse(null, { status: 404 })`
`return HttpResponse.json(deletedPost)`
how do you write the handler in Typescript?

`http.delete<{id: string}, ???, ???>(...)`?
Since this is an example about `params`, I guess the request body type should be `{}` or `undefined`?

What if the `id` was passed through json body instead of url parameter?

**Comment by mattcosta7** [comment]
https://www.typescriptlang.org/play?ts=5.3.2#code/PTAEGcCcGNgWwPbQNbmACwIYDsAmAbAU0nADoArcAKAEs4AHBSAF1AG9R1nn6AaUABLd6AJULhG2cIX4BlZpBrRmYiQimFQAX1AAzSAjigA5HHAB3Y1VBVmAT3qaACgnCsAvOy1Vo6t6Ex8fBc3cFBPbEJzUABZTHoAHjdFbABzfhDmAD4ACgBKaypCAA9GFlBfKVYsPCIScNAAbWtOYVJcQiJmQhzjYEZQ4AAuGlxjfhyOekxITDNtPPCs9hbQUBBQAEEg0EhCAEcAV3FWaeZ0UGnZ+ZnNegMAN1HCXFAabFBzzQAiK7nwb6rdZgGapQ5wQjYVgIXSfdCaPZqDS7cQIfAPYikIGVfwcUbaBp-MxAmiwnL2Rwwt6vACE7k8xmS71SxkWe2Yh0gHyEPFUkmkpGIBkg+UKazWGwAMoRmIyAtxCAxWMwEKBUrMAEZwu6uVgauxvZhhACSABEseKKn5WB0ui9Mg1AsFdWRUjKcqMCjZLRs+epXuYaOcAqBvgAWAAMYdAADkEKwAGIIQ54b4opHSN6wr5qmgY7BAjYDVhm0C4BDiUDYeOgEo0NwW8Wk0A5Gm2mX23WLNhAtbszkfSLRHmicT8nrYQ5BfgcNyYDngIagSPRrRe8Xeb3ijamzod7WXXV6AxGHPfJ2ZAGgODxRtrC8u9p77oe3BewtgP14UCB4OYUMAEwRhGoAAPIANJpoi44BN+Obtt0rzFneKIclygjCF+AqUOoOQIZ2bjrmuvAtFwPBPnaCRsKMS5MmkWi5H0xZoCMYwTFMMz-AsS4jlhmgAD6gPIijKHxCSZMs7jLD2PpgNs+AokcJyXPOFxEmEtyXI8zyvO8B6-JxZiArJASQGCEJQqAVI5tBfgIqi6KYti1rsNSBKeOpJJkhShBUvidIMnRLJsjKA4YbyY52YKkDCqKW4SmA0qyhpCpKp8qrqpgWo5sWoD6oaJrms5VRls+BEeAEQSXqQbrMK+74mV+AZBhc-7hlGsY1kmKa4FBkXIs2OapHmkIfoe-iluWlbVqwdYNl5LZtmVuCZN2vaoWFQ7haOGYTlO+AzhAzDzoci7Lh1a5ApuY27naB65fohj6Q+oRpje9AoS9RoUR29Vitun79c1f6AcBYGQemME4K88HLeNzAof26G8f12HgLh+ErV2LTEaRbT4VRNFHSkqQMb0-QusMozjC2HHXGEa48ZhqMCUJChKCoLPibqknSWN8mKcc-hnGphkaXsWkIE8HS6R8Z7qcZAOmeZkLQtm8KQ3Z6ZohikAoTirB4q8OgeWLC3kg4vmwv59ImEFrIbcjzO7dFsUNUrSVyvO3RpSqaqavdR75UGhX6y5mMOp4X2uu6nr-QloBNT+LUhu10ZxomyapprA3q5ow35mNuWTRWYQzbWxT1gj5tLXaWOESslp9qF6FbSju05JO06uXOC5LiuCxXfHwKgLd+45Uej2nhr55VS6b23kC0c-S+cfxSPSe-q1oMgRBfW7bBMMa5j8OIy33LO+OFDo9geHLatON5CRAC6QA

any of these work in typescript.   There are many other options too, so it's hard to answer that directly since there's multiple solutions to the same problems in the type system.  

Generally, I'd suggest providing a few generics as you can manually and letting the type inference work, and only add them when necessary

**Comment by kettanaito** [comment, marked as answer]
Hi, @Toshinaki. This is a great question!

Please refer to the [Using with TypeScript](https://mswjs.io/docs/best-practices/typescript) best practice page from now on. You can find examples on how to annotate path params, request and response bodies in [this section](https://mswjs.io/docs/best-practices/typescript#http-handlers).
